src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/poptimizer/simple/ReduceVocRewritingJoinPlanOptimizerImpl.java

```diff
@@ -0,0 +1,144 @@
+ * After determining join ordering, the physical plan can be constructed using bind join.
+ */
+public class ReduceVocRewritingJoinPlanOptimizerImpl extends CardinalityBasedJoinPlanOptimizerBase
+{
+    public ReduceVocRewritingJoinPlanOptimizerImpl(final FederationAccessManager fedAccessMgr ) {
+        super(fedAccessMgr);
+    }
+
+    @Override
+    public EnumerationAlgorithm initializeEnumerationAlgorithm( final List<PhysicalPlan> subplans ) {
+
+        return new CardinalityBasedJoinPlanOptimizerBase.GreedyConstructionAlgorithm(subplans) {
+            @Override
+            protected PhysicalPlan determineJoinOrderAndConstructPlan( final PhysicalPlanWithStatistics firstSubPlan, final List<PhysicalPlanWithStatistics> subPlansWithStatistics, final Comparator<PhysicalPlanWithStatistics> orderBasedOnCard) {
+                final PriorityQueue<PhysicalPlanWithStatistics> orderedCandidateSubPlans = new PriorityQueue<>(orderBasedOnCard);
+                PhysicalPlanWithStatistics nextSelectedSubPlan = firstSubPlan;
+                PhysicalPlan currentJoinPlan = nextSelectedSubPlan.plan;
+
+                while ( !subPlansWithStatistics.isEmpty() ){
+                    // Identify subplans that have join variables with the selected subplan as new candidateSubPlans.
+                    List<PhysicalPlanWithStatistics> tmpCandidateSubPlans = getSubPlansContainVars( currentJoinPlan.getExpectedVariables(), subPlansWithStatistics );
+                    if( tmpCandidateSubPlans.isEmpty() ){
+                        tmpCandidateSubPlans = subPlansWithStatistics;
+                    }
+
+                    // Keep candidate SubPlans that use the same vocabulary mapping with the previous selected SubPlan.
+                    // If 'candidateSubPlans' is empty, use all subplans in 'tmpCandidateSubPlans' as candidate subplans.
+                    List<PhysicalPlanWithStatistics> candidateSubPlans = getSubPlansWithSameVoc( nextSelectedSubPlan.getVocabularyMappings(), tmpCandidateSubPlans);
+                    if( candidateSubPlans.isEmpty() ){
+                        candidateSubPlans = tmpCandidateSubPlans;
+                    }
+
+                    // The candidateSubPlans are added to the ordered list 'orderedCandidateSubPlans'
+                    orderedCandidateSubPlans.addAll(candidateSubPlans);
+
+                    // select the first element (with the lowest cardinality) from orderedCandidateSubPlans as the nextSelectedSubPlan
+                    // Remove the selected subPlan from subPlansWithStatistics and clear orderedCandidateSubPlans for the next loop
+                    nextSelectedSubPlan = orderedCandidateSubPlans.poll();
+                    subPlansWithStatistics.remove(nextSelectedSubPlan);
+                    orderedCandidateSubPlans.clear();
+
+                    // Construct a physical plan using bind join
+                    currentJoinPlan = constructPlanWithBJ( currentJoinPlan, nextSelectedSubPlan.plan );
```

Comment:
Related to my previous comment about calling `nextSelectedSubPlan` `previousSelectedSubPlan`, here it makes sense to call this program variable `nextSelectedSubPlan`. Therefore, for the sake of the readability of the code, I would use `nextSelectedSubPlan` as a `final` variable here inside the loop and, then, at the end of the loop assign `nextSelectedSubPlan` to `previousSelectedSubPlan`.